### PR TITLE
Plugin implementation

### PR DESCRIPTION
**Pull request by rvnh**
Implementation of core of the plugin.

Main functionality of OpenAPI parser consists of two steps:
1. Parse JSON file using ilib-loctool-json
2. Parse result of the first step (each resource is parsed separately) using ilib-loctool-ghfm to process eventual Markdown inside the strings

Result consist of markdown typed resources that are extracted to an xliff file.

Localization consists of three steps:
1. Parse JSON resources using ilib-loctool-ghfm
2. Localize markdown resources
3. Use localized markdown resources to localize initial JSON.

### CODE REVIEW

File: OpenAPIFileType.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+var JsonFileType = require('ilib-loctool-json/JsonFileType');
+var MarkdownFileType = require('ilib-loctool-ghfm/MarkdownFileType');
+var OpenAPIFile = require('./OpenAPIFile');
+var logger = log4js.getLogger('loctool.plugin.OpenAPIFileType');
+
+var OpenAPIFileType = function(project) {
+	this.type = 'openapi';
+	this.datatype = 'openapi';
+	this.extensions = ['.json'];
```
Comment (ehoogerbeets) [inline OpenAPIFileType.js:30]: Should support '.jsn' files as well for Windows

File: OpenAPIFileType.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+ * @returns {OpenAPIFile}
+ */
+OpenAPIFileType.prototype.newFile = function(path) {
+	logger.debug('Add new file ' + path);
+
+	return new OpenAPIFile({
+		project: this.project,
+		pathName: path,
+		type: this
+	});
```
Comment (ehoogerbeets) [inline OpenAPIFileType.js:166]: Add in the new second parameter, options and pass the sourceLocale and targetLocale options to the new file, in case we ever need read a localized openapi.json and extract the localized strings from it

File: package.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    "url": "https://github.com/iLib-js/ilib-loctool-openapi.git"
+  },
+  "scripts": {
+    "dist": "npm pack",
+    "test": "node test/testSuite.js",
+    "debug": "node --inspect-brk test/testSuite.js",
+    "clean": "git clean -f -d test"
+  },
+  "engines": {
+    "node": ">=6.0"
```
Comment (ehoogerbeets) [inline package.json:49]: Should be >= 8.0 now... 6.0 doesn't allow us to do plugins properly

File: package.json [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  },
+  "dependencies": {
+    "ilib": "^14.8.0",
```
Comment (ehoogerbeets) [inline package.json:52]: Please do `npm update` to get the latest dependencies

File: test/testOpenAPIFile.js [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+			test.equal(actual, expected);
+
+			test.done();
+		}
+	},
+
+	localize: {
+		testLocalize: function(test) {
+			test.expect(0);
```
Comment (ehoogerbeets) [inline test/testOpenAPIFile.js:295]: Need more tests here

File: test/testOpenAPIFile.js [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	id: 'foo',
+	sourceLocale: 'en-US'
+}, './test/testfiles', {
+	locales:['ru-RU'],
+	targetDir: '.',
+	nopseudo: true,
+	openapi: {
+		schemas: ['./test/testfiles/openapi-schema.json'],
+		mappings: {
+
```
Comment (ehoogerbeets) [inline test/testOpenAPIFile.js:27]: Without mappings, we don't apply the json schema to the file. Instead, the JsonFile instance applies the generic schema that is tested below.

File: test/testOpenAPIFile.js [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+			var oaf = new OpenAPIFile({
+				project: p,
+				type: t
+			});
+			test.ok(oaf);
+
+			var jsonToParse = '{\n' +
+					'  "string1": "This is a test description"\n' +
+					'}';
```
Comment (ehoogerbeets) [inline test/testOpenAPIFile.js:102]: Need to have some tests of actual openapi.json snippets, including some tests of snippets that have markdown embedded in them.

File: test/testOpenAPIFileType.js [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+	getLocalizedPath: {
+		testGetLocalizedPathLocaleDir: function(test) {
+			test.expect(2);
+
+			var oaft = new OpenAPIFileType(p);
+
+			test.ok(oaft);
+
+			test.equals(
+					oaft.getLocalizedPath('resources/[localeDir]/openapi.json', 'dirA/dirB/openapi.json', 'de-DE'),
```
Comment (ehoogerbeets) [inline test/testOpenAPIFileType.js:78]: Please use 4 spaces instead of tab characters

File: test/testfiles/openapi-schema.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{
+  "$id": "openapi-schema",
+  "properties": {
+    "description": {
```
Comment (ehoogerbeets) [inline test/testfiles/openapi-schema.json:4]: Remove?

File: OpenAPIFileType.js (hunk 2 above)
Comment (rvnh) [reply OpenAPIFileType.js]: Please check if the change is what you expected. Also note that I've added assignment of `targetLocale` in `OpenAPIFile.js` L:32

File: test/testfiles/openapi-schema.json (hunk 9 above)
Comment (rvnh) [reply test/testfiles/openapi-schema.json]: Replaced it with a custom schema file to test possibility to use other than default schema to parse openapi files.

File: test/testOpenAPIFile.js (hunk 7 above)
Comment (rvnh) [reply test/testOpenAPIFile.js]: Replaced random JSON with parts of real OpenAPI json structure everywhere.

File: test/testOpenAPIFile.js (hunk 6 above)
Comment (rvnh) [reply test/testOpenAPIFile.js]: You're right! It was using default schema from json plugin. Added correct mapping and verified in tests that it's using OpenAPI schema now.

File: test/testOpenAPIFile.js (hunk 5 above)
Comment (rvnh) [reply test/testOpenAPIFile.js]: Added a test for localize method consisting of three steps:
1. Reading file and extracting strings.
2. Localizing it using mock `TranslationSet`
3. Checked the output resource file to be exactly as expected.

Also tested a couple of corner cases in this test, like string that was split into two due to a double new line and only one half has translation, resulting in half-translated string in the resource file.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    "debug": "node --inspect-brk test/testSuite.js",
+    "clean": "git clean -f -d test"
+  },
+  "engines": {
+    "node": ">=8.0"
+  },
+  "dependencies": {
+    "ilib": "^14.9.0",
+    "ilib-loctool-ghfm": "^1.7.1",
+    "ilib-loctool-json": "^1.1.0",
```
Comment (ehoogerbeets) [inline package.json:54]: This needs to be `^v1.2.0`. I have published your json changes to npm now.

**Comment by ehoogerbeets** [comment]
Just fix the json dependency and I can approve this!

**Review by ehoogerbeets — APPROVED** [review]
